presto-main/src/main/java/com/facebook/presto/memory/SynchronizedAggregatedMemoryContext.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+import com.google.common.io.Closer;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.io.IOException;
+
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public final class SynchronizedAggregatedMemoryContext
+        extends AbstractAggregatedMemoryContext
+        implements AutoCloseable
```

Comment:
I think you mixed two responsibilities here:
 - synchronize wrapper
 - ownership aware factory